src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -0,0 +1,537 @@
+
+    auto log_rcvbuf =
+        static_cast<int>(zeek::id::find_val<zeek::IntVal>("Cluster::Backend::ZeroMQ::log_rcvbuf")->AsInt());
+
+    ZEROMQ_DEBUG("Setting log_sndhwm=%d log_sndbuf=%d log_rcvhwm=%d log_rcvbuf=%d linger_ms=%d", log_sndhwm, log_sndbuf,
+                 log_rcvhwm, log_rcvbuf, linger_ms);
+
+    log_push.set(zmq::sockopt::sndhwm, log_sndhwm);
+    log_push.set(zmq::sockopt::sndbuf, log_sndbuf);
+    log_push.set(zmq::sockopt::linger, linger_ms);
+    log_push.set(zmq::sockopt::immediate, log_immediate);
+
+    log_pull.set(zmq::sockopt::rcvhwm, log_rcvhwm);
+    log_pull.set(zmq::sockopt::rcvbuf, log_rcvbuf);
+
+
+    if ( ! listen_log_endpoint.empty() ) {
+        ZEROMQ_DEBUG("Listening on log pull socket: %s", listen_log_endpoint.c_str());
+        try {
+            log_pull.bind(listen_log_endpoint);
+        } catch ( zmq::error_t& err ) {
+            zeek::reporter->Error("ZeroMQ: Failed to bind to PULL socket %s: %s", listen_log_endpoint.c_str(),
+                                  err.what());
+            return false;
+        }
+    }
+
+    const auto& log_endpoints = zeek::id::find_val<zeek::VectorVal>("Cluster::Backend::ZeroMQ::connect_log_endpoints");
+    for ( unsigned int i = 0; i < log_endpoints->Size(); i++ )
+        connect_log_endpoints.push_back(log_endpoints->StringValAt(i)->ToStdString());
+
+    for ( const auto& endp : connect_log_endpoints ) {
+        ZEROMQ_DEBUG("Connecting log_push socket with %s", endp.c_str());
+        try {
+            log_push.connect(endp);
+        } catch ( zmq::error_t& err ) {
+            zeek::reporter->Error("ZeroMQ: Failed to connect to PUSH socket %s: %s", endp.c_str(), err.what());
+            return false;
+        }
+    }
+
+    // We may not be connected yet, but there's no logging either unless
+    // we hook into zmq_socket_monitor.
+    //
+    // Check the following if we wanted to add ZeroMQ into the IO loop without
+    // the flare. Maybe?
+    //
+    // https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/
+    self_thread = std::thread(self_thread_fun, this);
+
+    // After connecting, call ThreadedBackend::DoInit() to register
+    // the IO source with the loop.
+    return ThreadedBackend::DoInit();
+}
+
+bool ZeroMQBackend::SpawnZmqProxyThread() {
+    proxy_thread = std::make_unique<ProxyThread>(listen_xpub_endpoint, listen_xsub_endpoint);
+    return proxy_thread->Start();
+}
+
+bool ZeroMQBackend::DoPublishEvent(const std::string& topic, const std::string& format,
+                                   const cluster::detail::byte_buffer& buf) {
+    // Parts to send for an event publish.
+    std::array<zmq::const_buffer, 4> parts = {
+        zmq::const_buffer(topic.data(), topic.size()),
+        zmq::const_buffer(my_node_id.data(), my_node_id.size()),
+        zmq::const_buffer(format.data(), format.size()),
+        zmq::const_buffer(buf.data(), buf.size()),
+    };
```

Comment:
Are these four parts something you come up with here, or is this in reference to somewhere else? If the former, could we spell it out in a comment higher up, and if the latter, link to it? I'm asking because people will sooner or later wonder how we're serializing things with ZeroMQ and it would be nice not to have to dig into the code for it (memories of CAF and needing to hunt down its wire protocol :-). I had the same thought below for log writes.

Reply:
Fair, I added some comments what the motivation for the parts are and rewrote the log writes piece to also use an array with a bit of documentation.

> I'm asking because people will sooner or later wonder how we're serializing things with ZeroMQ

It's really just sending of a ZeroMQ multipart message adding some structure and guarantee that all or none of it all arrive. With NATS, AMQP, or MQTT one might use message headers instead.

I hope people aren't overly interested in the serialization (outside of maybe implementing non-Zeek loggers). For integration with Zeek scripts, WebSockets should be the answer.